internal/characters/barbara/abil.go

```diff
@@ -0,0 +1,197 @@
+package barbara
+
+import (
+	"fmt"
+
+	"github.com/genshinsim/gcsim/pkg/core"
+)
+
+// Standard attack function with seal handling
+func (c *char) Attack(p map[string]int) (int, int) {
+	f, a := c.ActionFrames(core.ActionAttack, p)
+
+	ai := core.AttackInfo{
+		ActorIndex: c.Index,
+		Abil:       fmt.Sprintf("Normal %v", c.NormalCounter),
+		AttackTag:  core.AttackTagNormal,
+		ICDTag:     core.ICDTagNormalAttack,
+		ICDGroup:   core.ICDGroupDefault,
+		Element:    core.Hydro,
+		Durability: 25,
+		Mult:       attack[c.NormalCounter][c.TalentLvlAttack()],
+	}
+	done := false
+	// Taken from Noelle code
+	cb := func(a core.AttackCB) {
+		if done { //why do we need this @srl
+			return
+		}
+		//check for healing
+		if c.Core.Status.Duration("barbskill") > 0 {
+			//heal target
+			heal := (prochpp[c.TalentLvlSkill()] + prochp[c.TalentLvlSkill()])
+			c.Core.Health.HealAll(c.Index, heal)
+			done = true
+		}
+
+	}
+	c.Core.Combat.QueueAttack(ai, core.NewDefCircHit(0.1, false, core.TargettableEnemy), 0, f, cb)
+	c.AdvanceNormalIndex()
+
+	// return animation cd
+	return f, a
+}
+
+// Charge attack function - handles seal use
+func (c *char) ChargeAttack(p map[string]int) (int, int) {
+
+	f, a := c.ActionFrames(core.ActionCharge, p)
+
+	ai := core.AttackInfo{
+		ActorIndex: c.Index,
+		Abil:       "Charge Attack",
+		AttackTag:  core.AttackTagExtra,
+		ICDTag:     core.ICDTagNone,
+		ICDGroup:   core.ICDGroupDefault,
+		StrikeType: core.StrikeTypeBlunt,
+		Element:    core.Hydro,
+		Durability: 25,
+		Mult:       charge[c.NormalCounter][c.TalentLvlAttack()],
+	}
+
+	done := false
+	// Taken from Noelle code
+	cb := func(a core.AttackCB) {
+		if done { //why do we need this @srl
+			return
+		}
+		//check for healing
+		if c.Core.Status.Duration("barbskill") > 0 {
+			//heal target
+			heal := (prochpp[c.TalentLvlSkill()] + prochp[c.TalentLvlSkill()])
+			c.Core.Health.HealAll(c.Index, 4*heal)
+			done = true
+		}
+
+	}
+	var cbenergy func(a core.AttackCB) = nil
+	energyCount := 0
+	if c.Base.Cons >= 4 {
+		cbenergy = func(a core.AttackCB) {
+			//check for healing
+			if c.Core.Status.Duration("barbskill") > 0 && energyCount < 5 {
+				//regen energy
+				c.AddEnergy(1)
+				energyCount++
+			}
+
+		}
+	}
+
+	// TODO: Not sure of snapshot timing
+	c.Core.Combat.QueueAttack(ai, core.NewDefCircHit(2, false, core.TargettableEnemy), 0, f, cb, cbenergy)
+
+	return f, a
+}
+
+// barbara skill - copied from bennett burst
+
+func (c *char) Skill(p map[string]int) (int, int) {
```

Comment:
so she should have another on collision mechanic with her E that applies hydro to targets it touches similar to xingqiu

unfortunately i don't have the exact details on that mechanics. for now it's ok to leave it out i guess till i have more time to research it

Reply:
from Barbara mains:

Finding:
Barbara's melody loop ticks every 1.5 seconds and follows the ~universal rule of being able to apply its element (hydro) every 2.5s per target, and also each third hit within those 2.5s windows. The extra two damaging droplet strikes that occur immediately after activation count towards those three hits for an early, extra application of hydro, after which it will only have the capability to apply hydro every 3 seconds (every other tick).

Reply:
It looks like 4 hits at the beginning to me with only 2 of the hits registering doing damage and applying Hydro. https://streamable.com/bbesgl ... going to need instruction here, I might as well try to do something as close as I can.